zerver/lib/actions.py

```diff
@@ -572,24 +572,26 @@ def do_create_user(email: str, password: Optional[str], realm: Realm, full_name:
                    realm_creation: bool=False,
                    acting_user: Optional[UserProfile]=None) -> UserProfile:
 
-    user_profile = create_user(email=email, password=password, realm=realm,
-                               full_name=full_name,
-                               role=role, bot_type=bot_type, bot_owner=bot_owner,
-                               tos_version=tos_version, timezone=timezone, avatar_source=avatar_source,
-                               default_sending_stream=default_sending_stream,
-                               default_events_register_stream=default_events_register_stream,
-                               default_all_public_streams=default_all_public_streams,
-                               source_profile=source_profile)
+    with transaction.atomic():
+        user_profile = create_user(email=email, password=password, realm=realm,
+                                   full_name=full_name,
+                                   role=role, bot_type=bot_type, bot_owner=bot_owner,
+                                   tos_version=tos_version, timezone=timezone, avatar_source=avatar_source,
+                                   default_sending_stream=default_sending_stream,
+                                   default_events_register_stream=default_events_register_stream,
+                                   default_all_public_streams=default_all_public_streams,
+                                   source_profile=source_profile)
+
+        event_time = user_profile.date_joined
+        if not acting_user:
+            acting_user = user_profile
+        RealmAuditLog.objects.create(
+            realm=user_profile.realm, acting_user=acting_user, modified_user=user_profile,
+            event_type=RealmAuditLog.USER_CREATED, event_time=event_time,
+            extra_data=orjson.dumps({
+                RealmAuditLog.ROLE_COUNT: realm_user_count_by_role(user_profile.realm),
+            }).decode())
```

Comment:
Hmm, do you think `create_user(` itself should have the `@transaction.atomic` decorator, and the `RealmAuditLog` block moved into that function?  I think that'd be the cleaner API.

(Maybe even including `do_increment_logging_stat` and `update_license_ledger_if_needed`, since that's part of the database work?)

Hmm, or maybe what we actually want is to instead have this whole `do_create_user` in `@transaction.atomic`, and use `transaction.on_commit()` both here and in the `send_event` calls from within `process_new_human_user`?

That is likely the most correct implementation, but would have the same testing problems we're still working on elsewhere.

Reply:
@showell what are your thoughts?

Reply:
> Hmm, do you think `create_user(` itself should have the `@transaction.atomic` decorator, and the `RealmAuditLog` block moved into that function? I think that'd be the cleaner API.
> 
> (Maybe even including `do_increment_logging_stat` and `update_license_ledger_if_needed`, since that's part of the database work?)

That seems pretty uncontroversial to me, we just need to answer whether the two main places that call `create_user` explicitly want to run that additional logic - so `do_delete_user` and `create_mirror_user_if_needed`.





> Hmm, or maybe what we actually want is to instead have this whole `do_create_user` in `@transaction.atomic`, and use `transaction.on_commit()` both here and in the `send_event` calls from within `process_new_human_user`?
> 
> That is likely the most correct implementation, but would have the same testing problems we're still working on elsewhere. Hmm, and also would have trouble with `bulk_add_subs_to_db_with_logging` already having `transaction.atomic`.

Yeah that change would probably go in the #16721 PR. But there is some trouble here due to `process_new_human_user` doing a bunch of extra things - it sends events in more than one place (it sends messages in multiple places for example, which generates additional events), and calls `queue_json_publish` - all things that don't go well with atomic transactions and probably all require being registered in `.on_commit` - which gets pretty difficult to do and adds significant complexity.